src/coreclr/jit/rangecheck.h

```diff
@@ -72,6 +72,14 @@ static bool IntAddOverflows(int max1, int max2)
     return false;
 }
 
+static bool IntMultiplyOverflows(int max1, int max2)
+{
+    if (max1 == 0 || max2 == 0)
+        return false;
+    int res = max1 * max2;
+    return !(max1 == res / max2);
```

Comment:
This has undefined behavior, e.g. Clang optimizes this to `return false;`: https://godbolt.org/z/r9KE9r9xT
You should be able to use `CheckedOps::MulOverflows` in its place.
Can you add a test that fails when this function does not return the correct result?

Reply:
Thanks for catching that --- I've gone ahead and switched to `CheckedOps::MulOverflows` and added in a new test case, `ArrBoundsElim.cs/csproj`. I wasn't quite sure where to place it in the testing infrastructure, happy to rename/move it.

I found a case that will cause a problem if the overflow check behaves incorrectly: https://github.com/dotnet/runtime/blob/162ed95ad8c72ec44dea8e897c4bbe043933a0c8/src/tests/JIT/opt/AssertionPropagation/ArrBoundElim.cs#L75-L89

Reply:
Thanks, it's great that you were able to find a test case for this. However this allocates 28GB of memory, and I'm not sure that we can get a byte array of 2 GB reliably in our testing infrastructure -- is it possible to decrease the size of the array by a few magnitudes?

Reply:
Yes, I was able to find another case that will work with much lower array size and loop bounds that will trigger the same overflow condition. ~256mb array now.